pkg/pillar/cmd/zfsmanager/handlediskconfig.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2022 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zfsmanager
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	libzfs "github.com/bicomsystems/go-libzfs"
+	"github.com/lf-edge/eve/api/go/info"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/vault"
+	"github.com/lf-edge/eve/pkg/pillar/zfs"
+)
+
+func getVdevNameStatus(dev libzfs.VDevTree, name string) (string, libzfs.VDevStat) {
+	if dev.Type != libzfs.VDevTypeDisk {
+		for _, d := range dev.Devices {
+			retDev, retStat := getVdevNameStatus(d, name)
+			if retDev != "" {
+				return retDev, retStat
+			}
+		}
+	} else {
+		// we can receive partition number
+		devName, err := types.GetDiskNameByPartName(dev.Name)
+		if err != nil {
+			log.Errorf("failed to get disk name by part name")
+		}
+		if strings.Contains(devName, name) {
+			return devName, dev.Stat
+		}
+	}
+	return "", libzfs.VDevStat{}
+}
+
+func handleDiskConfigRestart(ctxArg interface{}, restartCounter int) {
+	log.Functionf("handleDiskConfigRestart(%d)", restartCounter)
+	ctx := ctxArg.(*zfsContext)
+	select {
+	case ctx.disksProcessingTrigger <- struct{}{}:
+	default:
+		log.Functionf("handleDiskConfigRestart: disksProcessingTrigger already triggered")
+	}
+}
+
+func processDisksTask(ctx *zfsContext) {
+
+	disksProcessingTicker := time.NewTicker(disksProcessingInterval)
+
+	wdName := agentName + "diskstask"
+	stillRunning := time.NewTicker(stillRunningInterval)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+
+	for {
+		select {
+		case <-disksProcessingTicker.C:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-ctx.disksProcessingTrigger:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func processDisks(ctx *zfsContext) error {
+	if vault.ReadPersistType() != types.PersistZFS {
+		return nil
+	}
+	disksConfig := ctx.subDiskConfig.GetAll()
+	if len(disksConfig) == 0 {
+		return nil
+	}
+	persistPool, err := libzfs.PoolOpen(vault.DefaultZpool)
+	if err != nil {
+		return fmt.Errorf("cannot open pool: %s", err)
+	}
+	defer persistPool.Close()
+	vdevTree, err := persistPool.VDevTree()
+	if err != nil {
+		return fmt.Errorf("cannot get vdev tree: %s", err)
+	}
+	for _, cfg := range ctx.subDiskConfig.GetAll() {
```

Comment:
Not sure how expensive is GetAll() call. You can as well use disksConfig calculated at line 78.

Reply:
Modified